basil/HL/bronkhorst_elflow.py

```diff
@@ -16,160 +17,100 @@
-        if self.debug != 0:
-            logger.debug("ELFLOW.write() %s" % str(cmd_s))
-        self._intf.write(cmd_s)
+        if time.time() - self.pre_time < 1.0:
+            time.sleep(1.0)
+        self._intf.write(str(cmd))
+        self.pre_time = time.time()
 
     def read(self):
-        ret_s = self._intf.read()
-        if self.debug != 0:
-            logger.debug("ELFLOW.read() %s" % str(ret_s))
-        if len(ret_s) < 5 or ret_s[0] != ":" or ret_s[3:5] != self.node:
-            logger.debug("ELFLOW.read() format error ret=%s" % str(ret_s))
-            return []
-        ret_len = int(ret_s[1:3])
-        if ret_len * 2 != len(ret_s[3:-2]):
-            logger.debug("ELFLOW.read() data lenth error ret=%s" % str(ret_s))
-            return []
-        ret = []
-        for i in range(ret_len - 1):
-            ret.append(int(ret_s[5 + 2 * i:5 + 2 * (i + 1)], 16))
-        return ret
+        ret = self._intf.read()
+        if len(ret) < 2 or ret[-2:] != "\r\n":
+            logger.warning("read() termination error")
+        return ret.strip()
+
+    def get_valve_output(self):
+        self._intf.write(self.CMDS['get_valve'])
+        ret = int(self.read()[11:], 16)
+        return ret * 61.7 / 10345949  # converts int in percentage
 
     def set_setpoint(self, value):
-        cmd = [1, 1, 0x21, (value >> 8) & 0xFF, value & 0xFF]
-        self.write(cmd)
+        """value range from 0 - 32000
+        """
+
+        if not isinstance(value, int):
+            raise ValueError(f"Given value has to be of type integer, is {type(value)}!")
+
+        hex_val = hex(value)[2:]        # [2:] to remove the 0x from the beginning of the hex number
+        command = f"{self.CMDS['set_setpoint']}" + f"{hex_val.zfill(4)}"  # hex should have at least 4 digits
+        self._intf.write(command)
         ret = self.read()
-        if len(ret) != 3:
-            logger.debug("ELFLOW.set_setpoint() data lenth error ret=%s" % str(ret))
-            return -1
-        elif ret[0] == 0 and ret[1] == 0 and ret[2] == 5:
-            return 0
-        else:
-            logger.debug("ELFLOW.set_setpoint() ret error ret=%s" % str(ret))
-            return -1
+        return ret
 
     def get_setpoint(self):
-        cmd = [4, 1, 0x21, 1, 0x21]
-        self.write(cmd)
+        self._intf.write(self.CMDS['get_setpoint'])
         ret = self.read()
-        if len(ret) != 5:
-            logger.debug("ELFLOW.set_setpoint() data lenth error ret=%s" % str(ret))
-            return -1
-        elif ret[0] == 2 and ret[1] == cmd[1] and ret[2] == cmd[2]:
-            return ((ret[3] << 8) & 0xFF00) | (ret[4] & 0xFF)
-        else:
-            logger.debug("ELFLOW.set_setpoint() ret error ret=%s" % str(ret))
-            return -1
-
-    def set_control_mode(self, value):
+        answer_in_hex = ret[11:]  # read from the 11th digits to translate what point is set
+        answer = int(answer_in_hex, 16)
+        return answer
```

Comment:
I think this might even be less "confusing" if you do `answer = int(ret[11:], 16)  # convert reply from hex to integer`

Reply:
This is changed!